File: src/backend/python/api/services/user_service.py
import hashlib
import time
from abc import ABC, abstractmethod
from datetime import datetime, timezone
from urllib.parse import quote

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from shared.config import settings
from shared.models.email_log import EmailLog
from shared.models.user import User
from api.repositories.user_repository import IUserRepository, UserRepository
from api.schemas.todo_item import PaginatedResponse
from api.schemas.user import (
    ChangePasswordRequest, ConfirmPasswordResetRequest, CreateUserRequest, ResetPasswordRequest,
    SignUpRequest, UpdateProfileRequest, UpdateUserRequest, UserResponse,
)
from api.security import create_signed_token, decode_signed_token, hash_password, verify_password


class IUserService(ABC):
    @abstractmethod
    def get_all(self, page: int = 1, page_size: int = 20) -> PaginatedResponse[UserResponse]: ...
    @abstractmethod
    def get_by_id(self, user_id: int) -> UserResponse: ...
    @abstractmethod
    def create(self, request: CreateUserRequest) -> UserResponse: ...
    @abstractmethod
    def update(self, user_id: int, request: UpdateUserRequest) -> UserResponse: ...
    @abstractmethod
    def set_active(self, user_id: int, is_active: bool) -> UserResponse: ...
    @abstractmethod
    def signup(self, request: SignUpRequest) -> UserResponse: ...
    @abstractmethod
    def get_profile(self, user_id: int) -> UserResponse: ...
    @abstractmethod
    def update_profile(self, user_id: int, request: UpdateProfileRequest) -> UserResponse: ...
    @abstractmethod
    def change_password(self, user_id: int, request: ChangePasswordRequest) -> None: ...
    @abstractmethod
    def request_password_reset(self, request: ResetPasswordRequest) -> None: ...
    @abstractmethod
    def confirm_password_reset(self, request: ConfirmPasswordResetRequest) -> None: ...


class UserService(IUserService):
    def __init__(self, repository: IUserRepository) -> None:
        self._repo = repository

    def _get_or_404(self, user_id: int) -> User:
        user = self._repo.get_by_id(user_id)
        if user is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"User {user_id} not found.")
        return user

    def _ensure_unique(self, username: str, email: str, excluding_id: int | None = None) -> None:
        if self._repo.username_exists(username.strip(), excluding_id):
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Username is already in use.")
        if self._repo.email_exists(email.strip().lower(), excluding_id):
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Email is already in use.")

    @staticmethod
    def _response(user: User) -> UserResponse:
        return UserResponse.model_validate(user)

    def get_all(self, page: int = 1, page_size: int = 20) -> PaginatedResponse[UserResponse]:
        page = max(1, page)
        page_size = min(settings.MAX_PAGE_SIZE, max(1, page_size))
        users, total = self._repo.get_all(skip=(page - 1) * page_size, limit=page_size)
        return PaginatedResponse(items=[self._response(user) for user in users], total=total, page=page,
                                 page_size=page_size, total_pages=-(-total // page_size))

    def get_by_id(self, user_id: int) -> UserResponse:
        return self._response(self._get_or_404(user_id))

    def create(self, request: CreateUserRequest) -> UserResponse:
        username, email = request.username.strip(), request.email.strip().lower()
        self._ensure_unique(username, email)
        user = User(username=username, email=email, password_hash=hash_password(request.password),
                    is_active=request.is_active, created_at=datetime.now(timezone.utc))
        return self._response(self._repo.add(user))

    def update(self, user_id: int, request: UpdateUserRequest) -> UserResponse:
        user = self._get_or_404(user_id)
        username = request.username.strip() if request.username is not None else user.username
        email = request.email.strip().lower() if request.email is not None else user.email
        self._ensure_unique(username, email, user_id)
        user.username, user.email = username, email
        if request.password is not None:
            user.password_hash = hash_password(request.password)
        user.updated_at = datetime.now(timezone.utc)
        return self._response(self._repo.update(user))

    def set_active(self, user_id: int, is_active: bool) -> UserResponse:
        user = self._get_or_404(user_id)
        user.is_active = is_active
        user.updated_at = datetime.now(timezone.utc)
        return self._response(self._repo.update(user))

    def signup(self, request: SignUpRequest) -> UserResponse:
        return self.create(CreateUserRequest(username=request.username, email=request.email,
                                             password=request.password, is_active=True))

    def get_profile(self, user_id: int) -> UserResponse:
        return self.get_by_id(user_id)

    def update_profile(self, user_id: int, request: UpdateProfileRequest) -> UserResponse:
        return self.update(user_id, UpdateUserRequest(username=request.username, email=request.email))

    def change_password(self, user_id: int, request: ChangePasswordRequest) -> None:
        user = self._get_or_404(user_id)
        if not user.is_active:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="The user account is inactive.")
        if not verify_password(request.current_password, user.password_hash):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="The current password is incorrect.")
        user.password_hash = hash_password(request.new_password)
        user.updated_at = datetime.now(timezone.utc)
        self._repo.update(user)

    def request_password_reset(self, request: ResetPasswordRequest) -> None:
        user = self._repo.get_by_email(request.email.strip().lower())
        if user is None or not user.is_active:
            return
        password_fingerprint = hashlib.sha256(user.password_hash.encode()).hexdigest()
        token = create_signed_token({"sub": str(user.id), "exp": int(time.time()) + settings.PASSWORD_RESET_TOKEN_LIFETIME_MINUTES * 60,
                                     "password": password_fingerprint}, settings.PASSWORD_RESET_SECRET_KEY)
        separator = "&" if "?" in settings.PASSWORD_RESET_CONFIRMATION_URL else "?"
        reset_url = f"{settings.PASSWORD_RESET_CONFIRMATION_URL}{separator}token={quote(token)}"
        self._repo.add_email_log(EmailLog(recipient=user.email, subject="Reset your Todo API password",
            body=f"Use this link to reset your password: {reset_url}\n\nThis link expires in {settings.PASSWORD_RESET_TOKEN_LIFETIME_MINUTES} minutes.",
            status="pending", created_at=datetime.now(timezone.utc)))

    def confirm_password_reset(self, request: ConfirmPasswordResetRequest) -> None:
        try:
            payload = decode_signed_token(request.token, settings.PASSWORD_RESET_SECRET_KEY)
            user = self._repo.get_by_id(int(payload["sub"]))
            fingerprint = hashlib.sha256(user.password_hash.encode()).hexdigest() if user else ""
            if user is None or not user.is_active or payload.get("password") != fingerprint:
                raise ValueError
        except (HTTPException, KeyError, TypeError, ValueError):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,
                                detail="The password reset token is invalid or expired.")
        user.password_hash = hash_password(request.new_password)
        user.updated_at = datetime.now(timezone.utc)
        self._repo.update(user)


def get_user_service(db: Session) -> IUserService:
    return UserService(UserRepository(db))

File: src/backend/go/internal/service/user_service.go
package service

import (
	"errors"
	"fmt"
	"math"
	"net/url"
	"strings"
	"time"

	"github.com/todo/backend/go/internal/config"
	"github.com/todo/backend/go/internal/dto"
	"github.com/todo/backend/go/internal/models"
	"github.com/todo/backend/go/internal/repository"
	"github.com/todo/backend/go/internal/security"
	"gorm.io/gorm"
)

var (
	ErrUserNotFound              = errors.New("user not found")
	ErrUsernameConflict          = errors.New("username is already in use")
	ErrEmailConflict             = errors.New("email is already in use")
	ErrInvalidCurrentPassword    = errors.New("the current password is incorrect")
	ErrInactiveUser              = errors.New("the user account is inactive")
	ErrInvalidPasswordResetToken = errors.New("the password reset token is invalid or expired")
	ErrInvalidCredentials        = errors.New("invalid email or password")
)

type UserService interface {
	GetAll(page, pageSize int) (dto.PaginatedResponse[dto.UserResponse], error)
	GetByID(id uint) (dto.UserResponse, error)
	Create(req dto.CreateUserRequest, actorUserID ...*uint) (dto.UserResponse, error)
	Update(id uint, req dto.UpdateUserRequest, actorUserID ...*uint) (dto.UserResponse, error)
	SetActive(id uint, active bool, actorUserID ...*uint) (dto.UserResponse, error)
	SignUp(req dto.SignUpRequest) (dto.UserResponse, error)
	GetProfile(id uint) (dto.UserResponse, error)
	UpdateProfile(id uint, req dto.UpdateProfileRequest) (dto.UserResponse, error)
	ChangePassword(id uint, req dto.ChangePasswordRequest) error
	RequestPasswordReset(req dto.ResetPasswordRequest) error
	ConfirmPasswordReset(req dto.ConfirmPasswordResetRequest) error
	CreateToken(req dto.TokenRequest) (dto.TokenResponse, error)
}

type userService struct {
	repo              repository.UserRepository
	cfg               *config.Config
	dummyPasswordHash string
}

func NewUserService(repo repository.UserRepository, cfg *config.Config) UserService {
	dummyHash, _ := security.HashPassword("not-a-real-password", cfg.PasswordHashIterations)
	return &userService{repo: repo, cfg: cfg, dummyPasswordHash: dummyHash}
}

func userResponse(u *models.User) dto.UserResponse {
	return dto.UserResponse{ID: u.ID, Username: u.Username, Email: u.Email, IsActive: u.IsActive,
		CreatedAt: u.CreatedAt, CreatedByUserID: u.CreatedByUserID,
		UpdatedAt: u.UpdatedAt, UpdatedByUserID: u.UpdatedByUserID}
}
func (s *userService) get(id uint) (*models.User, error) {
	u, err := s.repo.FindByID(id)
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, ErrUserNotFound
	}
	return u, err
}
func (s *userService) unique(username, email string, excluding *uint) error {
	exists, err := s.repo.UsernameExists(username, excluding)
	if err != nil {
		return err
	}
	if exists {
		return ErrUsernameConflict
	}
	exists, err = s.repo.EmailExists(email, excluding)
	if err != nil {
		return err
	}
	if exists {
		return ErrEmailConflict
	}
	return nil
}
func (s *userService) GetAll(page, size int) (dto.PaginatedResponse[dto.UserResponse], error) {
	if page < 1 {
		page = 1
	}
	if size < 1 {
		size = s.cfg.DefaultPageSize
		if size < 1 {
			size = 20
		}
	}
	if s.cfg.MaxPageSize > 0 && size > s.cfg.MaxPageSize {
		size = s.cfg.MaxPageSize
	}
	r, err := s.repo.FindAll((page-1)*size, size)
	if err != nil {
		return dto.PaginatedResponse[dto.UserResponse]{}, err
	}
	items := make([]dto.UserResponse, len(r.Items))
	for i := range r.Items {
		items[i] = userResponse(&r.Items[i])
	}
	return dto.PaginatedResponse[dto.UserResponse]{Items: items, Total: r.Total, Page: page, PageSize: size, TotalPages: int(math.Ceil(float64(r.Total) / float64(size)))}, nil
}
func (s *userService) GetByID(id uint) (dto.UserResponse, error) {
	u, e := s.get(id)
	if e != nil {
		return dto.UserResponse{}, e
	}
	return userResponse(u), nil
}
func (s *userService) Create(req dto.CreateUserRequest, actorUserID ...*uint) (dto.UserResponse, error) {
	username, email := strings.TrimSpace(req.Username), strings.ToLower(strings.TrimSpace(req.Email))
	if err := s.unique(username, email, nil); err != nil {
		return dto.UserResponse{}, err
	}
	hash, err := security.HashPassword(req.Password, s.cfg.PasswordHashIterations)
	if err != nil {
		return dto.UserResponse{}, err
	}
	active := true
	if req.IsActive != nil {
		active = *req.IsActive
	}
	u, err := s.repo.Create(&models.User{Username: username, Email: email, PasswordHash: hash,
		IsActive: active, CreatedByUserID: firstActor(actorUserID)})
	if err != nil {
		return dto.UserResponse{}, err
	}
	return userResponse(u), nil
}
func (s *userService) Update(id uint, req dto.UpdateUserRequest, actorUserID ...*uint) (dto.UserResponse, error) {
	u, err := s.get(id)
	if err != nil {
		return dto.UserResponse{}, err
	}
	username, email := u.Username, u.Email
	if req.Username != nil {
		username = strings.TrimSpace(*req.Username)
	}
	if req.Email != nil {
		email = strings.ToLower(strings.TrimSpace(*req.Email))
	}
	if err = s.unique(username, email, &id); err != nil {
		return dto.UserResponse{}, err
	}
	u.Username, u.Email = username, email
	if req.Password != nil {
		u.PasswordHash, err = security.HashPassword(*req.Password, s.cfg.PasswordHashIterations)
		if err != nil {
			return dto.UserResponse{}, err
		}
	}
	now := time.Now().UTC()
	u.UpdatedAt = &now
	u.UpdatedByUserID = firstActor(actorUserID)
	u, err = s.repo.Update(u)
	if err != nil {
		return dto.UserResponse{}, err
	}
	return userResponse(u), nil
}
func (s *userService) SetActive(id uint, active bool, actorUserID ...*uint) (dto.UserResponse, error) {
	u, e := s.get(id)
	if e != nil {
		return dto.UserResponse{}, e
	}
	u.IsActive = active
	now := time.Now().UTC()
	u.UpdatedAt = &now
	u.UpdatedByUserID = firstActor(actorUserID)
	u, e = s.repo.Update(u)
	if e != nil {
		return dto.UserResponse{}, e
	}
	return userResponse(u), nil
}
func (s *userService) SignUp(req dto.SignUpRequest) (dto.UserResponse, error) {
	active := true
	return s.Create(dto.CreateUserRequest{Username: req.Username, Email: req.Email, Password: req.Password, IsActive: &active})
}
func (s *userService) GetProfile(id uint) (dto.UserResponse, error) { return s.GetByID(id) }
func (s *userService) UpdateProfile(id uint, req dto.UpdateProfileRequest) (dto.UserResponse, error) {
	return s.Update(id, dto.UpdateUserRequest{Username: req.Username, Email: req.Email}, &id)
}
func (s *userService) ChangePassword(id uint, req dto.ChangePasswordRequest) error {
	u, e := s.get(id)
	if e != nil {
		return e
	}
	if !u.IsActive {
		return ErrInactiveUser
	}
	if !security.VerifyPassword(req.CurrentPassword, u.PasswordHash) {
		return ErrInvalidCurrentPassword
	}
	u.PasswordHash, e = security.HashPassword(req.NewPassword, s.cfg.PasswordHashIterations)
	if e != nil {
		return e
	}
	now := time.Now().UTC()
	u.UpdatedAt = &now
	u.UpdatedByUserID = &id
	_, e = s.repo.Update(u)
	return e
}
func (s *userService) RequestPasswordReset(req dto.ResetPasswordRequest) error {
	u, e := s.repo.FindByEmail(strings.ToLower(strings.TrimSpace(req.Email)))
	if errors.Is(e, gorm.ErrRecordNotFound) || u != nil && !u.IsActive {
		return nil
	}
	if e != nil {
		return e
	}
	token, e := security.CreateResetToken(u.ID, u.PasswordHash, s.cfg.PasswordResetSecretKey, time.Now().UTC().Add(time.Duration(s.cfg.PasswordResetTokenLifetimeMinutes)*time.Minute))
	if e != nil {
		return e
	}
	sep := "?"
	if strings.Contains(s.cfg.PasswordResetConfirmationURL, "?") {
		sep = "&"
	}
	resetURL := s.cfg.PasswordResetConfirmationURL + sep + "token=" + url.QueryEscape(token)
	_, e = s.repo.AddEmailLog(&models.EmailLog{Recipient: u.Email, Subject: "Reset your Todo API password", Body: fmt.Sprintf("Use this link to reset your password: %s\n\nThis link expires in %d minutes.", resetURL, s.cfg.PasswordResetTokenLifetimeMinutes), Status: "pending"})
	return e
}
func (s *userService) ConfirmPasswordReset(req dto.ConfirmPasswordResetRequest) error {
	id, fingerprint, e := security.DecodeResetToken(req.Token, s.cfg.PasswordResetSecretKey, time.Now().UTC())
	if e != nil {
		return ErrInvalidPasswordResetToken
	}
	u, e := s.get(id)
	if e != nil || !u.IsActive || security.PasswordFingerprint(u.PasswordHash) != fingerprint {
		return ErrInvalidPasswordResetToken
	}
	u.PasswordHash, e = security.HashPassword(req.NewPassword, s.cfg.PasswordHashIterations)
	if e != nil {
		return e
	}
	now := time.Now().UTC()
	u.UpdatedAt = &now
	u.UpdatedByUserID = nil
	_, e = s.repo.Update(u)
	return e
}

func (s *userService) CreateToken(req dto.TokenRequest) (dto.TokenResponse, error) {
	u, err := s.repo.FindByEmail(strings.ToLower(strings.TrimSpace(req.Email)))
	hash := s.dummyPasswordHash
	if err == nil && u != nil {
		hash = u.PasswordHash
	}
	valid := security.VerifyPassword(req.Password, hash)
	if errors.Is(err, gorm.ErrRecordNotFound) || err == nil && (u == nil || !valid || !u.IsActive) {
		return dto.TokenResponse{}, ErrInvalidCredentials
	}
	if err != nil {
		return dto.TokenResponse{}, err
	}
	now := time.Now().UTC()
	expiresIn := s.cfg.JWTTokenLifetimeMinutes * 60
	if expiresIn < 60 {
		expiresIn = 60
	}
	token, err := security.CreateJWT(u.ID, s.cfg.JWTSecretKey, now, now.Add(time.Duration(expiresIn)*time.Second))
	if err != nil {
		return dto.TokenResponse{}, err
	}
	return dto.TokenResponse{AccessToken: token, TokenType: "Bearer", ExpiresIn: expiresIn}, nil
}
